doc/design/embargo-mvp.md

```diff
@@ -0,0 +1,131 @@
+
+When releasing an embargoed dandiset, all asset data for that dandiset is copied to the public bucket.
+
+When uploading a new asset to an embargoed dandiset, the server will first check if that blob has already been uploaded publically.
+If so, the public blob will be used instead of uploading the data again to the embargo bucket.
+
+
+# Django changes
+
+## Models
+The `Dandiset` model will have an `embargo_status` field that is one of `PRIVATE`, `RELEASING`, or `PUBLIC`.
+* `PUBLIC` means that the Dandiset is publically accessible and publishable.
+  This is the state all Dandisets currently have.
+* `PRIVATE` means that the Dandiset is embargoed.
+  It is not searchable or viewable to non-owners.
+  It is not publishable.
+  Manifest YAML/JSON files will be written to the embargo bucket rather than the public bucket.
+* `RELEASING` means that the Dandiset is currently transitioning from embargoed to public.
+  All modification operations will return a 400 error while the release is in progress.
+  This includes metadata changes and uploads.
+
+A new `EmbargoedAssetBlob` model will be added.
+This behaves the same as a normal `AssetBlob`, but points to the embargo bucket rather than the public bucket.
+It also contains a reference to the `Dandiset` it belongs to.
+
+`Asset`s will point to exactly one `AssetBlob` or `EmbargoedAssetBlob`.
+An `Asset` will be considered "embargoed" if it has an `EmbargoedAssetBlob`.
+An embargoed `Asset` can only belong to the dandiset named in the `EmbargoedAssetBlob`.
+The automatically populated `Asset` metadata will also set the `access` field to the appropriate value depending on whether or not the asset is embargoed.
+
+The `Upload` model will have a new optional field `embargoed_dandiset` that points to an embargoed dandiset.
+If specified, the uploaded data will be sent to the embargoed bucket instead of the public bucket, and the `Upload` will create an `EmbargoedAssetBlob` when finalized.
+
+## API
+* Create dandiset endpoint
+
+  Add a new optional `award_number` field.
+  If set, the `access` metadata field will be set to `EmbargoedAccess` and the award number will be recorded as a `Funder` on a new Contributor Organization.
+  The `embargo_status` of the dandiset will also be set to `PRIVATE`.
+
+* Get/List dandiset endpoint
+
+  The DandisetViewSet queryset will filter out dandisets with `embargo_status == PRIVATE` that are also not owned by the current user.
+  This will prevent them from showing up in the listing or fetching endpoints.
+
+  The `DandisetSerializer` will also be updated to include the `embargo_status` field so that the web client can render the embargoed dandiset appropriately.
+
+* publish version endpoint
+
+  Return error 400 if `dandiset.embargo_status != PUBLIC`.
+
+* create asset, update metadata, and any other dandiset/version modification endpoints:
+
+  Return error 400 if `dandiset.embargo_status == RELEASING`.
```

Comment:
503 is almost certainly the wrong thing--that indicates that the whole server is (temporarily) borked. I don't think we need to send a Retry-After here; we can instead send a reason in the response body and allow the client to try again after a while.

Reply:
indeed -- 503 is no good, thank you @waxlamp for clarifying.  
I still think that some explicit, server provided "Retry-After" is better than some "a while" (a second? minute? ...) which might cause some eager clients retry way too often.  But it seems there is no other suitable response code with Retry-After (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After). And since we will not be monitoring individual connections etc, we can't use [429 "Too many requests"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429) really